user/baseline.man
.\" $Id: baseline.man,v 11.8 2025/09/28 23:02:39 tom Exp $
.TH BASELINE 1 2025-09-28 "" "User commands"
.ie n .ds CW R
.el   \{
.ie \n(.g .ds CW CR
.el       .ds CW CW
.\}
.de Es
.ne \\$1
.nr mE \\n(.f
.RS 5n
.sp .7
.nf
.nh
.ta 9n 17n 25n 33n 41n 49n
.ft \*(CW
..
.de Eh
.ft \\n(mE
.fi
.hy \\n(HY
.RE
.sp .7
..
.SH NAME
\fBbaseline\fR \-
RCS utility
.SH SYNOPSIS
\fBbaseline\fI [options] [file-specifications]\fR
.SH DESCRIPTION
\fBBaseline\fR permits you to set a baseline version in one or
more RCS files.
.PP
\fBBaseline\fR adds a baseline (placeholder) version to one or
more RCS archives.
The effective check-in time for the file is the
file's last check-in date (before \fBbaseline\fR is invoked)
.PP
Assuming that the file was originally checked in using \fBcheckin\fR,
then the check-in, modification and baseline dates all are
the same.
.PP
A baseline version consists of a group of files which
can be referenced by a common version number.
The baseline program uses for this purpose 2.0, 3.0, etc.
Version 1.0 is never used,
since the first development version is, by convention, 1.1.
.PP
After baselining, you may use \fBcheckout\fR to obtain
a particular set of files.
For example,
.Es
checkout \-r2.0 file
.Eh
.PP
obtains the baseline "2.0" version, while
.Es
checkout \-r2 file1 file2 file3
.Eh
.PP
obtains the last development version beginning with "2.0", but
before "3.0".
.SH OPTIONS
\fBBaseline\fR recognizes several options:
.TP 4n
.BI \- integer
\fRthe new baseline version, specified as a whole
number after a hyphen, e.g.,
.RS
.Es
baseline \-2 file1 file2 file3
.Eh
.PP
If no baseline version is specified, the most recent baseline
version (as recorded in the \fBRCS,v\fR \fBpermit\fR-file) is
assumed.
This is done to permit additional files to be added to an
existing baseline.
.RE
.TP
.B \-a
permits \fBbaseline\fR to process files in directories
beginning with ".".
.TP
.B \-l
leaves files locked after baselining.
Also, if files
are locked but not modified, no error is reported.
.TP
.B \-L
allows \fBbaseline\fR to follow symbolic links to directories.
.TP
.BI \-m msg
adds the string "\-\-\ \fImsg\fR" to
the default baseline revision message "\fBBASELINE\fR \fIdate\fR".
.TP
.B \-n
runs \fBbaseline\fR in no-op mode,
showing where permissions would be purged,
and listing files which would be baselined.
.TP
.B \-p
the "purge" option purges existing access lists in the archives
to prevent locks while the baselining process is proceeding.
.TP
.B \-R
permits \fBbaseline\fR to run recursively when
it encounters a non-RCS directory.
.TP
.B \-v
makes the baseline process more verbose by causing
permit, checkin and checkout to show their actions.
.SH OPERATIONS
Before invoking \fBbaseline\fR, ensure that all of the files to
be baselined have been checked into \fIrcs\fR
(e.g., with "\fBcheckin\fR \-u").
This is easily done using the directory editor \fBded\fR.
.PP
\fBBaseline\fR operates only upon files which have been
checked into \fIrcs\fR (i.e., a corresponding archive file exists
in the subdirectory "RCS").
.PP
\fBBaseline\fR requires that all of the working files have been
checked in (i.e., with the "\-u" option so that they are retained
as working files).
.PP
\fBBaseline\fR first invokes \fBpermit\fR (to ensure that you
have permission to place locks in the directory, then to purge the
archive access lists, and finally to set the baseline-version for
the directory).
It then invokes \fBcheckin\fR and \fBcheckout\fR
(which must be in your path) to first place a lock on the specified
files, and then to check-in the file with the new, baseline version.
.PP
If any files are found which are locked, \fBbaseline\fR prints
an error message and halts.
Similarly, an attempt made to baseline
a file with a lower version than has been archived will result in
an error.
.SH ENVIRONMENT
\fBBaseline\fR is written in C, and runs under POSIX systems.
.SH FILES
\fBBaseline\fR uses the following files
.TP
\fBpermit\fR
the RCS directory-permissions utility
.TP
\fBcheckin\fR
the RCS check-in package
.TP
\fBcheckout\fR
the RCS check-out package
.SH ANTICIPATED CHANGES
\fBBaseline\fR does not currently handle branch deltas.
.SH AUTHORS
Thomas E. Dickey <dickey@invisible-island.net>
.SH SEE ALSO
checkin, checkout, ded, permit


CHANGES
2025-09-28  Thomas E. Dickey  <dickey@invisible-island.net>

	* package/cm_tools-12.0.spec, package/debian/changelog: bump

	* package/debian/copyright: formatting

	* user/baseline.man, user/checkin.man, user/checkout.man,
	  user/checkup.man, user/copy.man, user/link2rcs.man, user/permit.man,
	  user/rcsput.man:
	check-manpage

	* user/baseline.man, user/checkin.man, user/checkout.man,
	  user/checkup.man, user/copy.man, user/link2rcs.man, user/permit.man,
	  user/rcsput.man:
	fixup-th-man

	* user/rcsget.man, user/vcs.man: check-manpage

	* user/rcsget.man, user/vcs.man: fixup-th-man

	* configure: regen

	* aclocal.m4, configure.in: use CF_PROG_INSTALL

2025-07-09  Thomas E. Dickey  <dickey@invisible-island.net>

	* config.guess, config.sub: update to 2025-07-10

2025-06-08  Thomas E. Dickey  <dickey@invisible-island.net>

	* install-sh: fix license

2025-01-17  Thomas E. Dickey  <dickey@invisible-island.net>

	* package/debian/copyright, COPYING, package/cm_tools-12.0.spec,
	  package/debian/changelog:
	bump

	* configure: regen

	* configure.in: use CF_WITH_SYSTYPE early

2025-01-06  Thomas E. Dickey  <dickey@invisible-island.net>

	* src/checkin/src/checkin.c, src/checkout/src/checkout.c,
	  src/checkup/src/checkup.c, src/checkout/src/rcsget.c,
	  src/copy/src/copy.c, src/link2rcs/src/link2rcs.c,
	  src/permit/src/permit.c, src/vcs/src/delete.c, src/vcs/src/unlock.c,
	  src/checkin/src/rcsput.c, src/baseline/src/baseline.c:
	gcc15-warnings

2024-12-21  Thomas E. Dickey  <dickey@invisible-island.net>

	* aclocal.m4: resync with my-autoconf

2024-07-28  Thomas E. Dickey  <dickey@invisible-island.net>

	* config.guess: 2024/07/27

2024-06-03  Thomas E. Dickey  <dickey@invisible-island.net>

	* config.sub: 2024/05/27

2024-04-25  Thomas E. Dickey  <dickey@invisible-island.net>

	* package/debian/copyright, COPYING, package/cm_tools-12.0.spec,
	  package/debian/changelog:
	bump

	* configure: regen

	* aclocal.m4: resync with my-autoconf

2023-12-27  Thomas E. Dickey  <dickey@invisible-island.net>

	* config.sub: 2023-12-27

2023-12-02  Thomas E. Dickey  <dickey@invisible-island.net>

	* config.guess: 2023-12-02

2023-01-22  Thomas E. Dickey  <dickey@invisible-island.net>

	* package/cm_tools-12.0.spec: bump

	* configure: regen

	* bin/makefile.in:
	build-fix for Solaris: rename variable $(USER) to $(AS_USER)

	* package/debian/copyright, COPYING: bump

	* aclocal.m4, configure.in: use pkg-config to find td_lib

	* package/debian/changelog: bump

2023-01-20  Thomas E. Dickey  <dickey@invisible-island.net>

	* package/debian/compat: bump

2022-12-31  Thomas E. Dickey  <dickey@invisible-island.net>

	* package/cm_tools-12.0.spec, package/debian/changelog: bump

	* aclocal.m4: build-fixes for NetBSD

	* configure: regen

2022-10-11  Thomas E. Dickey  <dickey@invisible-island.net>

	* user/makefile.in: use install-sh if needed

	* package/cm_tools-12.0.spec, package/debian/changelog,
	  package/debian/copyright, COPYING:
	bump

	* configure: regen

	* aclocal.m4: resync with my-autoconf

2022-08-01  Thomas E. Dickey  <dickey@invisible-island.net>

	* config.guess, config.sub: 2022-08-01

2021-03-28  Thomas E. Dickey  <dickey@invisible-island.net>

	* package/cm_tools-12.0.spec, package/debian/changelog: bump

	* configure: regen

	* aclocal.m4: resync with my-autoconf

2021-01-10  Thomas E. Dickey  <dickey@invisible-island.net>

	* configure: regen

	* aclocal.m4: quoting fixes for shellcheck

	* configure: regen

	* aclocal.m4: resync with my-autoconf

	* COPYING, package/cm_tools-12.0.spec, package/debian/changelog,
	  package/debian/copyright:
	bump

2020-12-08  Thomas E. Dickey  <dickey@invisible-island.net>

	* package/cm_tools-12.0.spec, package/debian/changelog: bump

	* user/makefile.in, src/vcs/src/makefile.in,
	  src/baseline/test/makefile.in, src/baseline/makefile.in,
	  src/checkin/test/makefile.in, src/checkout/src/makefile.in,
	  src/checkout/test/makefile.in, src/checkup/src/makefile.in,
	  src/checkup/test/makefile.in, src/copy/src/makefile.in,
	  src/copy/test/makefile.in, src/link2rcs/src/makefile.in,
	  src/link2rcs/test/makefile.in, support/makefile.in:
	use VPATH

2020-12-07  Thomas E. Dickey  <dickey@invisible-island.net>

	* src/checkin/makefile.in, src/checkout/makefile.in,
	  src/checkup/makefile.in, src/copy/makefile.in,
	  src/link2rcs/makefile.in, src/permit/makefile.in, src/vcs/makefile.in:
	use VPATH

	* package/debian/compat: bump

	* src/baseline/src/makefile.in, src/checkin/src/makefile.in,
	  src/permit/src/makefile.in, src/permit/test/makefile.in,
	  src/vcs/test/makefile.in:
	use VPATH

2020-12-06  Thomas E. Dickey  <dickey@invisible-island.net>

	* package/debian/copyright, COPYING: bump

	* configure: regen

	* aclocal.m4: resync with my-autoconf

	* configure.in: without-X

	* package/cm_tools-12.0.spec, package/debian/changelog: bump

2020-11-30  Thomas E. Dickey  <dickey@invisible-island.net>

	* install-sh: update to 2020-11-14

2020-08-16  Thomas E. Dickey  <dickey@invisible-island.net>

	* config.guess, config.sub: 2020-08-17

2019-12-21  Thomas E. Dickey  <dickey@invisible-island.net>

	* package/cm_tools-12.0.spec, package/debian/changelog: bump

	* user/checkup.man: fix typo

2019-12-06  Thomas E. Dickey  <dickey@invisible-island.net>

	* package/cm_tools-12.0.spec, package/debian/changelog: bump

	* user/rcsput.man, user/rcsget.man, user/permit.man, user/link2rcs.man,
	  user/checkup.man, user/checkin.man, user/copy.man, user/checkout.man,
	  user/baseline.man:
	check-manpage fixes (rename Ex to Eh for consistency, other errata)

2019-12-05  Thomas E. Dickey  <dickey@invisible-island.net>

	* package/cm_tools-12.0.spec, package/debian/changelog: bump

	* src/checkin/src/rcsput.c, src/checkin/src/checkin.c:
	use DYN-argv lists to replace catarg()

	* src/vcs/src/vcs.c, src/permit/src/permit.c:
	use show_argv2(), to handle verb vs argv[0]

	* src/checkout/src/checkout.c, src/checkout/src/rcsget.c:
	use DYN-argv lists to replace catarg()


src/baseline/src/baseline.c
/*
 * Title:	baseline.c (rcs baseline)
 * Author:	T.E.Dickey
 * Created:	24 Oct 1989
 * Modified:
 *		04 Dec 2019, use "executev()" and "show_argv()"
 *		22 Sep 1993, gcc warnings
 *		17 Jan 1992, added "-L" option.  Renamed "-r" to "-R" option
 *			     and "-f" to "-p" for consistency.
 *		22 Oct 1991, use 'shoarg()'
 *		11 Oct 1991, converted to ANSI
 *		20 May 1991, apollo sr10.3 cpp complains about endif-tags
 *		16 Apr 1990, "-f" option must allow 'permit' to run, otherwise
 *			     the RCS,v file won't be updated.  Made "-f" only
 *			     suppress 'permit' "-p" option.
 *		03 Jan 1990, added "-l" option for stluka. added logic (for "-l"
 *			     option) so that we can allow a locked file to be
 *			     baselined (if it is locked, but not changed).
 *			     Corrected code which infers baseline revision (if
 *			     no rcs directory existed, this did not notice!)
 *		07 Nov 1989, added "-f" option, tuned some verboseness.
 *			     Corrected treatment of "-r" option (recur only
 *			     when asked).  Corrected code which infers current-
 *			     baseline version.
 *		01 Nov 1989, walktree passes null-pointer to stat-block if
 *			     no-access.
 *
 * Function:	Use checkin/checkout to force a tree of files to have a common
 *		baseline-version number.
 *
 *		This is a rewrite of the Bourne-shell script 'baseline.sh'.
 *
 * Options:	(see 'usage()')
 */

#define	STR_PTYPES
#include	<ptypes.h>
#include	<rcsdefs.h>
#include	<sccsdefs.h>
#include	<ctype.h>
#include	<time.h>

MODULE_ID("$Id: baseline.c,v 11.11 2025/01/07 00:46:41 tom Exp $")

#define	isDIR(mode)	((mode & S_IFMT) == S_IFDIR)
#define	isFILE(mode)	((mode & S_IFMT) == S_IFREG)
#define	isLINK(mode)	((mode & S_IFMT) == S_IFLNK)

typedef struct _list {
    struct _list *link;
    char *text;
} LIST;

static time_t began;
static int revision = -1;
static char m_option[BUFSIZ];	/* message to show as "-m" */
static char r_option[80];

				/* options */
static int a_opt;		/* all-directory scan */
static int lock_up;		/* leave files locked */
static char *m_text;		/* message-text */
static int no_op;		/* no-op mode */
#ifdef	S_IFLNK
static int L_opt;
#endif
static int purge_opt;
static int recur;
static int verbose;

static void
doit(char **argv, int really)
{
    if (verbose) {
	show_argv(stdout, argv);
    }
    if (really) {
	if (executev(argv) < 0) {
	    FPRINTF(stderr, "?? %s\n", argv[0]);
	    exit(FAIL);
	}
    }
}

static void
purge_rights(const char *path)
{
    int argc;
    char *argv[20];
    char myrev[80];
    static LIST *purged;
    LIST *p;

    for (p = purged; p; p = p->link) {
	if (!strcmp(p->text, path))
	    return;
    }
    p = ALLOC(LIST, 1);
    p->link = purged;
    p->text = txtalloc(path);
    purged = p;

    argc = 0;
    argv[argc++] = txtalloc("permit");
    FORMAT(myrev, "-b%d", revision);
    argv[argc++] = myrev;
    if (!verbose)
	argv[argc++] = txtalloc("-q");
    if (no_op)
	argv[argc++] = txtalloc("-n");
    if (purge_opt)
	argv[argc++] = txtalloc("-p");
    argv[argc++] = m_option;
    argv[argc++] = txtalloc(rcs_dir(NULL, NULL));
    argv[argc] = NULL;
    doit(argv, no_op < 2);
}

static void
baseline(const char *path, const char *name, time_t edited)
{
    int argc;
    char *argv[20];
    char myrev[80];
    const char *version;
    const char *locker;
    time_t date;
    int cmp;
    int pending;		/* true if lock implies change */

    purge_rights(path);
    rcslast(path, name, &version, &date, &locker);
    pending = (!lock_up) || (date != edited);

    if (*version == '?') {
	if (istextfile(name)) {
	    FPRINTF(stderr, "?? %s (not archived)\n", name);
	    if (!recur)
		exit(FAIL);
	} else
	    PRINTF("** %s (ignored)\n", name);
	return;
    } else if ((cmp = vercmp(r_option, version, FALSE)) < 0) {
	FPRINTF(stderr, "?? %s (version mismatch -b%d vs %s)\n",
		name, revision, version);
	exit(FAIL);
	/*NOTREACHED */
    } else if (cmp == 0) {
	PRINTF("** %s (already baselined)\n", name);
	return;
    } else if (*locker != '?' && pending) {
	FPRINTF(stderr, "?? %s (locked by %s)\n", name, locker);
	exit(FAIL);
	/*NOTREACHED */
    }
    PRINTF("** %s\n", name);

    if (*locker == '?') {	/* wasn't locked, must do so now */
	argc = 0;
	argv[argc++] = txtalloc("checkout");
	argv[argc++] = txtalloc("-l");
	if (!verbose)
	    argv[argc++] = txtalloc("-q");
	argv[argc++] = txtalloc(name);
	argv[argc] = NULL;
	doit(argv, !no_op);
    }

    argc = 0;
    argv[argc++] = txtalloc("checkin");
    FORMAT(myrev, "-r%d.0", revision);
    argv[argc++] = myrev;
    argv[argc++] = txtalloc("-f");
    argv[argc++] = txtalloc("-u");
    if (!verbose)
	argv[argc++] = txtalloc("-q");
    if (lock_up)
	argv[argc++] = txtalloc("-l");
    argv[argc++] = m_option;
    argv[argc++] = txtalloc("-sRel");
    argv[argc++] = txtalloc(name);
    argv[argc] = NULL;
    doit(argv, !no_op);
}

static int
WALK_FUNC(scan_tree)
{
    char tmp[BUFSIZ];
    char *s = pathcat(tmp, path, name);

    if (sp == NULL || level > recur)
	readable = -1;
    else if (isDIR(sp->st_mode)) {
	abspath(s);		/* get rid of "." and ".." names */
	if (!a_opt && *pathleaf(s) == '.')
	    readable = -1;
	else if (sameleaf(s, sccs_dir(path, name))
		 || sameleaf(s, rcs_dir(path, name))) {
	    readable = -1;
	} else {
	    if (level > recur) {
		PRINTF("** %s (ignored)\n", name);
		readable = -1;
	    } else {
#ifdef	S_IFLNK
		if (!L_opt) {
		    struct stat sb;
		    if (lstat(name, &sb) < 0
			|| isLINK(sb.st_mode)) {
			PRINTF("** %s (link)\n", name);
			readable = -1;
		    }
		} else
#endif
		    track_wd(s);
	    }
	}
    } else if (isFILE(sp->st_mode)) {
	if (level > 0)
	    track_wd(path);
	baseline(path, name, sp->st_mtime);
    } else
	readable = -1;

    return (readable);
}

static void
do_arg(const char *name)
{
    int infer_rev = FALSE;

    FORMAT(m_option, "-mBASELINE %s", ctime(&began));
    m_option[strlen(m_option) - 1] = EOS;
    if (m_text)
	(void) strcat(strcat(m_option, " -- "), m_text);

    /*
     * If no revision was given, infer it, assuming that we are going to
     * add forgotten stuff to the current baseline.
     */
    if (revision < 0) {
	char vname[BUFSIZ];
	time_t date;
	const char *locker, *version;
	rcslast(".",
		vcs_file(rcs_dir(NULL, NULL), vname, FALSE),
		&version, &date, &locker);
	if (*version == '?'
	    || sscanf(version, "%d.", &revision) <= 0)
	    revision = 2;
	infer_rev = TRUE;
    }
    FORMAT(r_option, "%d.0", revision);

    (void) walktree((char *) 0, name, scan_tree, "r", 0);

    if (infer_rev)
	revision = -1;
}

static void
usage(void)
{
    static const char *tbl[] =
    {
	"Usage: baseline [options] [files]"
	,""
	,"Bumps versions of rcs files to the specified integer value (i.e., 2.0)."
	,"If no files are specified, all files in the current directory are assumed."
	,""
	,"Options are:"
	,"  -{integer}  baseline version (must be higher than last baseline)"
	,"  -a          process directories beginning with \".\""
	,"  -l          leave files locked after baselining"
	,"  -m TEXT     append reason for baseline to date-message"
#ifdef	S_IFLNK
	,"  -L          traverse symbolic links to directories"
#endif
	,"  -n          no-op (show what would be done, but don't do it)"
	,"  -p          run \"permit\" first to purge access lists"
	,"  -R          recur when lower-level directories are found"
	,"  -v          verbose (shows details)"
    };
    unsigned j;
    for (j = 0; j < sizeof(tbl) / sizeof(tbl[0]); j++)
	FPRINTF(stderr, "%s\n", tbl[j]);
    FFLUSH(stderr);
    exit(FAIL);
}

/*ARGSUSED*/
_MAIN
{
    int j;
    char *s;
    char *d;
    int had_args = FALSE;

    track_wd((char *) 0);
    began = time((time_t *) 0);

    for (j = 1; j < argc; j++) {
	if (*(s = argv[j]) == '-') {
	    while (*(++s)) {
		if (isdigit(UCH(*s))) {
		    if ((revision = (int) strtol(s, &d, 10)) < 2)
			usage();
		    s = d - 1;
		} else if (*s == 'm') {
		    if (*(++s)) {
			m_text = s;
			break;
		    } else
			usage();
		} else {
		    switch (*s) {
		    case 'a':
			a_opt++;
			break;
		    case 'l':
			lock_up++;
			break;
		    case 'n':
			no_op++;
			break;
#ifdef	S_IFLNK
		    case 'L':
			L_opt++;
			break;
#endif
		    case 'p':
			purge_opt++;
			break;
		    case 'R':
			recur = 999;
			break;
		    case 'v':
			verbose++;
			break;
		    default:
			usage();
		    }
		}
	    }
	} else {
	    do_arg(s);
	    had_args = TRUE;
	}
    }

    if (!had_args) {
	recur++;
	do_arg(".");
    }
    exit(SUCCESS);
    /*NOTREACHED */
}
